sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -418,21 +424,41 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
 
     if (DDLUtils.isDatasourceTable(withStatsProps)) {
-      val oldDef = client.getTable(db, withStatsProps.identifier.table)
-      // Sets the `schema`, `partitionColumnNames` and `bucketSpec` from the old table definition,
-      // to retain the spark specific format if it is. Also add old data source properties to table
-      // properties, to retain the data source table format.
-      val oldDataSourceProps = oldDef.properties.filter(_._1.startsWith(SPARK_SQL_PREFIX))
+      val oldTableDef = client.getTable(db, withStatsProps.identifier.table)
+
+      // Always update the location property w.r.t. the new table location.
+      val locationProp = tableDefinition.storage.locationUri.map { location =>
+        TABLE_LOCATION -> location
+      }
+      // Only update the `locationUri` field if the location is really changed, because this table
+      // may be not Hive-compatible and can not set the `locationUri` field. We should respect the
+      // old `locationUri` even it's None.
+      val oldLocation = getLocationFromRawTable(oldTableDef)
+      val locationUri = if (oldLocation == tableDefinition.storage.locationUri) {
```

Comment:
When we using in-memory catalog, the location is changed for `renameTable` on managed data source tables. 

It sounds like we are not having the same behavior when using hive external catalog?